spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -189,92 +204,136 @@ public Subscriber subscribe(String subscription,
+			public void onSuccess(Empty empty) {
+				settableListenableFuture.set(null);
+			}
+		});
+
+		return settableListenableFuture;
 	}
 
 	@Override
-	public void nack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
-
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::nack);
+	public ListenableFuture<Void> nack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+		return modifyAckDeadline(acknowledgeablePubsubMessages, 0);
 	}
 
 	@Override
-	public void modifyAckDeadline(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages,
-			int ackDeadlineSeconds) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
-		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds can't be less than 0.");
+	public ListenableFuture<Void> modifyAckDeadline(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages, int ackDeadlineSeconds) {
+		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be empty.");
+		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds must not be negative.");
+
+		AcknowledgeablePubsubMessage acknowledgeablePubsubMessage = acknowledgeablePubsubMessages.iterator().next();
+		ProjectSubscriptionName projectSubscriptionName = acknowledgeablePubsubMessage.getProjectSubscriptionName();
+		List<String> ackIds = collectAckIds(projectSubscriptionName, acknowledgeablePubsubMessages);
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages)
-				.forEach((sub, ackIds) -> modifyAckDeadline(sub, ackIds, ackDeadlineSeconds));
+		SettableListenableFuture<Void> settableListenableFuture = new SettableListenableFuture<>();
+
+		ApiFuture<Empty> apiFuture = modifyAckDeadline(
+				projectSubscriptionName.getSubscription(), ackIds, ackDeadlineSeconds);
+
+		ApiFutures.addCallback(apiFuture, new ApiFutureCallback<Empty>() {
+			@Override
+			public void onFailure(Throwable throwable) {
+				settableListenableFuture.setException(throwable);
+			}
+
+			@Override
+			public void onSuccess(Empty empty) {
+				settableListenableFuture.set(null);
+			}
+		});
+
+		return settableListenableFuture;
 	}
 
-	/**
-	 * Groups {@link AcknowledgeablePubsubMessage} messages by subscription.
-	 * @return a map from subscription to list of ack IDs.
-	 */
-	private Map<String, List<String>> groupAcknowledgeableMessages(
+	private List<String> collectAckIds(ProjectSubscriptionName projectSubscriptionName,
 			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		return acknowledgeablePubsubMessages.stream()
-				.collect(Collectors.groupingBy(AcknowledgeablePubsubMessage::getSubscriptionName,
-						Collectors.mapping(AcknowledgeablePubsubMessage::getAckId, Collectors.toList())));
+
+		List<String> ackIds = new LinkedList<>();
```

Comment:
Why not `ArrayList` initialized to the size of the `acknowledgeablePubsubMessages`?

Reply:
Fixed (no specific reason it wasn't set.)